### ISSUE TITLE
Add selection subgroups and a hotkey to cycle them

### ISSUE THREAD

**Issue by dragunoff** [OpenRA/OpenRA#16951]
This is an idea taken from StarCraft II and Warcraft 3 but slightly modified. The gist is that there is a hotkey (e.g. <kbd>Tab</kbd>) that cycles through the unit classes in the current selection. That creates a foreground (active) and a background (inactive) selection. It's best illustrated with this GIF that I made with a prototype implementation:
![OpenRA-CycleSelectionSubgroups](https://user-images.githubusercontent.com/1355810/63202961-18c90c80-c095-11e9-84e2-a39e1e6d77c1.gif)

The difference to the SC/WC implementation is that there all of the units remain active (i.e. receive orders), and to command only the subselection a modifier key is used. This however is not applicable to CNC games because modifiers are used for other things (force move/attack ground). For details on the SC/WC implementation see this video: https://www.youtube.com/watch?v=3vSV-sMX4f8&t=322s

Some remarks/ideas:
- This would be very useful for micro-ing armies as it allows command of different unit classes without losing the current selection. 
- In addition to the hotkey it can also be tied to a selection palette (see #16687)
- There could be hotkeys that cycles backwards (e.g. <kbd>Shift</kbd> + <kbd>Tab</kbd>) and brings back the whole selection to the foreground (e.g. <kbd>Ctrl</kbd> + <kbd>Tab</kbd>)

**Comment by pchote** [comment]
One could argue that effectively managing unit subgroups with e.g. control groups are part of the skill that separates the good players from the average ones, and part of what separates C&C from Starcraft.

I'm not sure where my opinion on that falls yet, but I do know that we're running really short on hotkeys for this - <kbd>tab</kbd> is already used for cycling production structures.

**Comment by dragunoff** [comment]
> 
> 
> One could argue that effectively managing unit subgroups with e.g. control groups are part of the skill that separates the good players from the average ones, and part of what separates C&C from Starcraft.
> 
Yes, **micro-management is a skill**. But one could also argue that **with the existing tools it's rather cumbersome to do it effectively**. Take the select-by-type hotkey for example - it's often times not useful because it selects units on the whole screen and with the small unit sizes in C&C games it's not uncommon to have more than one army on the screen. Yes, there is a zoom option that effectively limits the number of visible (i.e. selectable) units, so by combining zoom and select-by-type the player could go around that limitation. But on top of that the player has to also do a dance with the mouse - click on a tiny unit, select by type, give orders elsewhere on screen. One can argue that this is part of the skill... 

With regards to the comparison with StarCraft - I think unit micro is much more important there (smaller number of units, many spell-casters). And that's why SC2 / WC3 provide more advanced tools to perform micro management. Of course there is BroodWar - where they decided to keep the old UI constraints exactly because that's part of the game which should be mastered. It's an important debate especially when it comes to classic games. 

> I'm not sure where my opinion on that falls yet, but I do know that we're running really short on hotkeys for this - tab is already used for cycling production structures.
>
IMHO that hotkey is not very useful and should be unbound. I don't think I've ever seen anyone use it or mention it.

**Comment by pchote** [comment]
The main point there is that units are too small on the screen - this is a bug rather than a feature, as the renderer only supports integer scaling of things on the screen. #10382 is on my priority list for Next + 1, and will hopefully provide a better solution for this for most players.

**Comment by dragunoff** [comment]
The main point is that intra-blob micro-management is difficult. Zooming only gets around that problem (to some extent). A fair point against the proposed feature is that it may not fit the C&C games and I won't argue against that because it's a fine line... However in the long-term and thinking about OpenRA as an engine - this could be supported at least on the interface level so that mods can implement the logic.
